src/app/pages/my-request-detail/my-request-detail.component.ts

```diff
@@ -277,30 +244,58 @@ export class MyRequestDetailComponent implements OnInit {
               finalised: false,
               id: this.request?.id ?? '',
             };
-            return this.dataRequests.deleteDataElementMultiple(itemList, dataModel);
+            return this.dataRequests.deleteDataElementMultiple(items, dataModel);
           } else {
             return EMPTY;
           }
         })
       )
-      .subscribe((result: DataElementMultipleOperationResult) => {
-        const success = result.failures.length === 0;
-        let message = `${result.successes.length} Data element${
-          result.successes.length === 1 ? '' : 's'
-        } removed from request "${this.request?.label}".`;
-        if (!success) {
-          message += `\r\n${result.failures.length} Data element${
-            result.failures.length === 1 ? '' : 's'
-          } caused an error.`;
-          result.failures.forEach((item: DataElementOperationResult) =>
-            console.log(item.message)
-          );
+      .subscribe((resultMultiple: DataElementMultipleOperationResult) => {
+        const labels = resultMultiple.successes.map((success) => {
+          return success.item.label;
+        });
+
+        // Remove data element from request queries.
+        this.removeDataElementFromQuery(labels, this.dataQueryType);
```

Comment:
Each of these `removeDataElementFromQuery()` calls actually creates a nested `subscribe()` call too. This is considered an anti-pattern in Angular/RxJs because now the operations become non-deterministic, at worst producing race conditions. You can kind of see the effect a bit in the UI when it refreshes - the loading screen disappears once the element is removed, but maybe a second or two later the query components re-render when their operations complete.

It is normally better to have all these observables run together in the same `pipe()` and have one `subscribe()` at the end which handles the final success case. This makes everything deterministic, even if there are potential delays e.g. with network calls to complete. At least you can then guarantee that each steps runs in the exact order you expect.

A cut down version of this then might be:

```ts
removeOneOrMoreDataElements(items: DataElementSearchResult[]) {
  this.okCancelItemList(items)
    .afterClosed()
    .pipe(
      switchMap((result: OkCancelDialogResponse) => {
        // Show UI we're doing something...
        this.broadcast.loading({ loading: true });
        // Perform action to remove data element...
        return this.dataRequests.deleteDataElementMultiple(items, dataModel);        
      }),
      switchMap((resultMultiple: DataElementMultipleOperationResult) => {
        // Perform two async actions but wait for both to finish...
        return forkJoin([
          this.dataRequests.deleteDataElementsFromQuery(this.request.id, 'data', items),
          this.dataRequests.deleteDataElementsFromQuery(this.request.id, 'cohort', items),
        ])
      }),
      finalize(() => {
        // Always stop the "loading" overlay, even if an error occurs...
        this.broadcast.loading({ loading: true });
      })
    )
    .subscribe(() => {
      // Display success messages here...
    })
}
```

This probably won't compile straight up, and you may need to pass the results of each action down the pipe to be used later, but this would be the general idea.

Reply:
Refactored to solve antipattern, thanks